src/test/java/com/geeksforless/scenario/ScenarioServiceImplTest.java

```diff
@@ -0,0 +1,123 @@
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.security.core.Authentication;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class ScenarioServiceImplTest {
+
+    @Mock
+    private WorkerClient workerClient;
+
+    @Mock
+    private Authentication authentication;
+
+    @InjectMocks
+    private ScenarioServiceImpl scenarioService;
+
+    @BeforeEach
+    void setUp() {
+        CustomUserDetails userDetails = new CustomUserDetails();
+        userDetails.setId(1L);
+        userDetails.setEmail("user@example.com");
+        userDetails.setPassword("password");
+        lenient().when(authentication.getPrincipal()).thenReturn(userDetails);
+    }
+
+    @Test
+    void testSendScenario() {
+        ScenarioFormDTO formData = new ScenarioFormDTO("[]", false);
+        ScenarioRequestDTO expectedRequest = ScenarioRequestDTO.builder()
+                .userId(1L)
+                .scenarioList(Collections.emptyList())
+                .proxyRequired(false)
+                .build();
+        scenarioService.sendScenario(formData, authentication);
+        verify(workerClient).postScenarios(expectedRequest);
+        verifyNoMoreInteractions(workerClient);
+    }
+
+    @Test
+    void testGetScenarioResults() {
+        Long userId = 1L;
+        List<ScenarioResultDTO> expectedResults = Collections.singletonList(new ScenarioResultDTO());
+        when(workerClient.getScenariosByUserId(userId)).thenReturn(expectedResults);
+        List<ScenarioResultDTO> actualResults = scenarioService.getScenarioResults(authentication);
+        assertEquals(expectedResults, actualResults);
+    }
+
+    @Test
+    void testGetScenarioById() {
+        Integer scenarioId = 1;
+        ScenarioResultDTO expectedResult = new ScenarioResultDTO();
+        when(workerClient.getScenarioById(scenarioId)).thenReturn(expectedResult);
+        ScenarioResultDTO actualResult = scenarioService.getScenarioById(scenarioId);
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void testSendScenario_ValidJson_ShouldPostScenarios() {
+        String validJson = "[{\"name\":\"Test scenario\",\"site\":\"example.com\"," +
+                "\"steps\":[{\"action\":\"Click\",\"value\":\"Button\"}]}]";
+        ScenarioFormDTO formData = new ScenarioFormDTO(validJson, true);
+
+        scenarioService.sendScenario(formData, authentication);
+
+        ScenarioRequestDTO expectedRequest = ScenarioRequestDTO.builder()
+                .userId(1L)
+                .scenarioList(Collections.singletonList(new Scenario("Test scenario", "example.com",
+                        Collections.singletonList(new Step("Click", "Button")))))
+                .proxyRequired(true)
+                .build();
```

Comment:
what's difference between this test and the *testSendScenario*?

Reply:
The main difference is what data is used to create the ScenarioFormDTO and expectedRequest objects, and whether the postScenarios method is called from workerClient depending on the input data.